include/swift/Localization/LocalizationFormat.h

```diff
@@ -155,6 +155,7 @@ class SerializedLocalizationWriter {
 
 class LocalizationProducer {
   bool printDiagnosticName;
+  std::string localizedDebugDiagnosticMessage;
```

Comment:
This string will allocate a copy of the `localizedMessage.str() + diagnosticName.str()` from `getMessageOr`. Is this correct?

Reply:
No, this doesn't work because subsequent diagnostic requests (even notes for the current diagnostic) would replace this data and deallocate previous storage. We need to get new allocator added to `DiagnosticEngine` to retain diagnostic strings for the duration of compiler invocation or switch `getMessageOr` to return `std::string`.

Reply:
Thank you so much for explaining, this makes total sense. The reason I didn't go with switching `getMessageOr` to return a `std::string` is because I think it's quite an easy fix and I'm more interested in learning how to do the other approach. 

Based on my current understanding of your explanation, I think we need some sort of container to store a copy of all of the diagnostic messages for the duration of compiler invocation. Does that make sense? If not, I think I'll need more clarification and I'm afraid I don't know what "new allocator" means because as I understand I think it means some sort of variable that allocates something.

Reply:
You could to add a new field of type `llvm::BumpPtrAllocator` for `DiagnosticEngine` and pass it down to localization producers, they'd use it to allocate copies of strings to be returned by diagnostic engine if printing of the names is requested. Another idea - since "print diagnostic names" is a flag we could actually record all localizations with their respective names already while loading localizations, right?

Reply:
Thank you for bearing with me on this @xedin! So, now I think what I should do is:

1. Add an `llvm::BumpPtrAllocator` and `llvm::StringSaver` fields in `DiagnosticEngine`.
2. Pass the `llvm::StringSaver`  to the `LocalizationProducer`.
3. In `getMessageOr` I'll `save` the `localizedDebugDiagnosticMessage` in the `llvm::StringSaver`.

I think the code in `getMessageOr` will be something like:
```cpp
auto localizedDebugDiagnosticMessage = localizationSaver.save(localizedMessage.str() + diagnosticName.str());
```

Does that make sense to you?

Reply:
That sounds like a plan to me!

Reply:
@xedin I really miss the project! I'll try to implement this on my weekend, which is on Friday and Saturday. :)

Reply:
No rush!